validation/validate.py

```diff
@@ -0,0 +1,215 @@
+import pytest
+import re
+from typing import Any
+
+
+def read_yaml_file(file_name: str) -> str:
+    '''
+    Reads a YAML file and returns it as a dictionary.
+    '''
+    with open(file_name, 'r') as f:
+        try:
+            d = yaml.safe_load(f)
+        except yaml.YAMLError as e:
+            print(e)
+    return d
+
+
+def type_matches(value: Any, expected_type: str) -> bool:
+    '''
+    Checks whether value has the type expected_type.
+    If yes, returns True, otherwise False.
+
+    Allowed types T are: 'str', 'int', 'float', 'complex', 'bool',
+                         as well as 'List[T]'.
+
+    For any other types the function raises ValueError.
+    '''
+    allowed_basic_types = ['str', 'int', 'float', 'complex', 'bool']
+    allowed_list_types = ['List[{}]'.format(t) for t in allowed_basic_types]
+    allowed_types = allowed_basic_types + allowed_list_types
+
+    # first verify whether expected_type is allowed
+    if expected_type not in allowed_types:
+        raise ValueError('could not recognize expected_type: {}'.format(expected_type))
+
+    expected_type_is_list = re.match(r"^List\[\w+\]$", expected_type) is not None
+
+    if expected_type_is_list:
+        # make sure that value is actually a list
+        if not type(value).__name__ == 'list':
+            return False
+
+        # iterate over each element of the list
+        # and check whether it matches T
+        list_element_type = re.search(r"^List\[(\w+)\]$", expected_type).group(1)
+        for element in value:
+            if not type(element).__name__ == list_element_type:
+                return False
+
+        return True
+    else:
+        # expected type is a basic type
+        # this is the simpler case
+        return type(value).__name__ == expected_type
+
+
+def test_type_matches():
+
+    # str
+    assert type_matches('a string', 'str')
+    assert not type_matches(1.0, 'str')
+
+    # float and int
+    assert type_matches(1.0, 'float')
+    assert type_matches(1.0e-8, 'float')
+    assert type_matches(1, 'int')
+    assert not type_matches(1, 'float')
+
+    # complex
+    assert type_matches(1.0+1.0j, 'complex')
+
+    # bool
+    assert type_matches(True, 'bool')
+    assert type_matches(False, 'bool')
+    assert not type_matches(0, 'bool')
+
+    # lists
+    assert type_matches([1, 2, 3], 'List[int]')
+    assert not type_matches((1, 2, 3), 'List[int]')
+    assert not type_matches([1, 2, 3], 'List[float]')
+    assert type_matches([1.0, 2.0, 3.0], 'List[float]')
+    assert not type_matches([1.0, 2.0, 3], 'List[float]')
+
+    # unexpected type input
+    with pytest.raises(ValueError):
+        assert type_matches('example', 'weird')
+        assert type_matches('example', 'List[strange]')
+
+
+class InputError(Exception):
+    pass
+
+
+class TemplateError(Exception):
+    pass
+
+
+def validate_node(input_dict, template_dict):
+    '''
+    Verify a node. A node is either the root of the input or it is a section.
```

Comment:
For the docstrings, we should adhere to the NumPy style guide: https://numpydoc.readthedocs.io/en/latest/format.html

Reply:
I agree, but we can adjust this after the code is in.

Reply:
What I meant by this: the code will still change quite a bit, probably even in this PR so I did not want to make the docstrings too beautiful.

Reply:
Totally understand, but I'd prefer if the un-WIP form of this PR had docstrings in the "right" format :smiling_imp:

Reply:
OK - first rough stone masonry. Later cosmetics.